File: familybox/apu/apu.py
"""NES APU (Audio Processing Unit) core implementation.

Coordinates four audio channels (2 pulse, 1 triangle, 1 noise) using
a frame counter for timing, and mixes their output using the standard
NES mixing formula for final audio output.
"""
#  *
#  * @Author: ShaoqiLiang
#  * @Date: 2026-05-16 22:07:49
#  * @LastEditors: ShaoqiLiang
#  *

from familybox.apu.noise import NoiseChannel
from familybox.apu.pulse import PulseChannel
from familybox.apu.triangle import TriangleChannel

# NTSC CPU frequency (cycles per second)
_CPU_FREQ: int = 1_789_773

# Frame counter step thresholds (in CPU cycles)
# 4-step mode: quarter frames at 7457, 14913, 22371, 29829
_FOUR_STEP_QUARTER: list[int] = [7457, 14913, 22371, 29829]
# 5-step mode: quarter frames at 7457, 14913, 22371, 29829, 37281
_FIVE_STEP_QUARTER: list[int] = [7457, 14913, 22371, 29829, 37281]


class APU:
    """NES Audio Processing Unit.

    Manages four audio channels and provides mixed audio output.
    Uses a frame counter to schedule envelope, length counter,
    and sweep updates at quarter-frame and half-frame rates.
    """

    def __init__(self) -> None:
        self._pulse1: PulseChannel = PulseChannel()
        self._pulse2: PulseChannel = PulseChannel()
        self._triangle: TriangleChannel = TriangleChannel()
        self._noise: NoiseChannel = NoiseChannel()

        # Frame counter state
        self._frame_counter: int = 0
        self._frame_mode: int = 0  # 0 = 4-step, 1 = 5-step
        self._frame_step: int = 0
        self._irq_inhibit: bool = False

        # Audio output
        self._sample_buffer: list[float] = []
        self._sample_timer: float = 0.0
        self._sample_rate: int = 44100
        self._cycles_per_sample: float = _CPU_FREQ / self._sample_rate

    def reset(self) -> None:
        """Reset APU and all channels to initial state."""
        self._pulse1.reset()
        self._pulse2.reset()
        self._triangle.reset()
        self._noise.reset()
        self._frame_counter = 0
        self._frame_step = 0
        self._sample_buffer.clear()
        self._sample_timer = 0.0

    def write_register(self, addr: int, value: int) -> None:
        """Route register write to the appropriate channel.

        Args:
            addr: APU register address ($4000-$4017).
            value: 8-bit value to write.
        """
        if 0x4000 <= addr <= 0x4003:
            self._pulse1.write_register(addr, value)
        elif 0x4004 <= addr <= 0x4007:
            self._pulse2.write_register(addr, value)
        elif 0x4008 <= addr <= 0x400B:
            self._triangle.write_register(addr, value)
        elif 0x400C <= addr <= 0x400F:
            self._noise.write_register(addr, value)
        elif addr == 0x4015:
            self._write_status(value)
        elif addr == 0x4017:
            self._frame_mode = (value >> 7) & 0x01
            self._irq_inhibit = bool(value & 0x40)
            self._frame_counter = 0
            self._frame_step = 0

    def tick(self, cycles: int = 1) -> float:
        """Advance APU by *cycles* CPU cycles.

        Returns:
            Mixed audio sample (last computed, or 0.0 if no sample produced).
        """
        # Advance frame counter and check for envelope/length updates
        self._frame_counter += cycles
        self._check_frame_steps()

        # Batch tick channels (advance timers by N cycles)
        self._batch_tick_channels(cycles)

        # Sample rate conversion: mix only when a sample is needed
        sample = 0.0
        self._sample_timer += cycles
        if self._sample_timer >= self._cycles_per_sample:
            self._sample_timer -= self._cycles_per_sample
            p1 = self._pulse1.output()
            p2 = self._pulse2.output()
            t = self._triangle.output()
            n = self._noise.output()
            sample = self._mix(p1, p2, t, n)
            self._sample_buffer.append(sample)

        return sample

    def _batch_tick_channels(self, cycles: int) -> None:
        """Advance all channel timers by *cycles* at once."""
        # Pulse channels (two instances, same logic)
        for pulse in (self._pulse1, self._pulse2):
            ps = pulse._state
            if ps.timer >= cycles:
                ps.timer -= cycles
            else:
                remaining = cycles
                while remaining > 0:
                    if ps.timer < remaining:
                        remaining -= ps.timer + 1
                        ps.timer = ps.timer_period
                        pulse._duty_index = (pulse._duty_index + 1) & 0x07
                    else:
                        ps.timer -= remaining
                        break

        # Triangle channel
        tri = self._triangle
        ts = tri._state
        if ts.timer >= cycles:
            ts.timer -= cycles
        else:
            remaining = cycles
            while remaining > 0:
                if ts.timer < remaining:
                    remaining -= ts.timer + 1
                    ts.timer = ts.timer_period
                    if ts.length_counter > 0 and tri._linear_counter > 0:
                        ts.sequence_index = (ts.sequence_index + 1) & 0x1F
                else:
                    ts.timer -= remaining
                    break

        # Noise channel — precompute LFSR steps to reduce function calls
        noise = self._noise
        ns = noise._state
        if ns.timer >= cycles:
            ns.timer -= cycles
        else:
            period = ns.timer_period
            if period == 0:
                ns.timer = 0
            else:
                lfsr_steps = 0
                remaining = cycles
                while remaining > 0:
                    if ns.timer < remaining:
                        remaining -= ns.timer + 1
                        ns.timer = period
                        lfsr_steps += 1
                    else:
                        ns.timer -= remaining
                        break
                if lfsr_steps:
                    noise._tick_lfsr_n(lfsr_steps)

    def _check_frame_steps(self) -> None:
        """Check if the frame counter has crossed any step thresholds."""
        if self._frame_mode == 0:
            steps = _FOUR_STEP_QUARTER
            max_steps = 4
            half_set = frozenset((1, 3))
        else:
            steps = _FIVE_STEP_QUARTER
            max_steps = 5
            half_set = frozenset((0, 2))

        while (
            self._frame_step < max_steps
            and self._frame_counter >= steps[self._frame_step]
        ):
            self._tick_frame(self._frame_step in half_set)
            self._frame_step += 1
            if self._frame_step >= max_steps:
                self._frame_counter = 0
                self._frame_step = 0

    def _tick_frame(self, half_frame: bool) -> None:
        """Process a frame step.

        Args:
            half_frame: If True, also update length counters and sweep.
        """
        # Quarter frame: always update envelope and linear counter
        self._pulse1.tick_quarter_frame()
        self._pulse2.tick_quarter_frame()
        self._triangle.tick_quarter_frame()
        self._noise.tick_quarter_frame()

        # Half frame: additionally update length counters and sweep
        if half_frame:
            self._pulse1.tick_half_frame()
            self._pulse2.tick_half_frame()
            self._triangle.tick_half_frame()
            self._noise.tick_half_frame()

    def _write_status(self, value: int) -> None:
        """Handle write to $4015 (status register).

        Enables/disables individual channels and resets length counters
        when a channel is disabled.
        """
        self._pulse1.state.enabled = bool(value & 0x01)
        self._pulse2.state.enabled = bool(value & 0x02)
        self._triangle.state.enabled = bool(value & 0x04)
        self._noise.state.enabled = bool(value & 0x08)

        # Reset length counters when channels are disabled
        if not (value & 0x01):
            self._pulse1.state.length_counter = 0
        if not (value & 0x02):
            self._pulse2.state.length_counter = 0
        if not (value & 0x04):
            self._triangle.state.length_counter = 0
        if not (value & 0x08):
            self._noise.state.length_counter = 0

    @staticmethod
    def _mix(p1: int, p2: int, t: int, n: int) -> float:
        """Mix channel outputs using the NES mixing formula.

        The NES uses two separate nonlinear mixing stages:
        - Pulse mix: combines two pulse channels
        - TND mix: combines triangle, noise, and DMC (DMC not implemented)

        Args:
            p1: Pulse 1 output (0-15).
            p2: Pulse 2 output (0-15).
            t: Triangle output (0-15).
            n: Noise output (0-15).

        Returns:
            Mixed sample in range [0.0, ~1.0].
        """
        # Pulse mixing
        pulse_sum = p1 + p2
        pulse_out = 0.0
        if pulse_sum > 0:
            pulse_out = 95.88 / ((8128.0 / pulse_sum) + 100.0)

        # TND mixing (triangle + noise + DMC)
        tnd_out = 0.0
        tnd_val = t / 8227.0 + n / 12241.0
        if tnd_val > 0.0:
            tnd_out = 159.79 / ((1.0 / tnd_val) + 100.0)

        return pulse_out + tnd_out

    def get_sample_buffer(self) -> list[float]:
        """Get and clear the audio sample buffer.

        Returns:
            List of mixed audio samples since last call.
        """
        buffer = self._sample_buffer
        self._sample_buffer = []
        return buffer

    @property
    def pulse1(self) -> PulseChannel:
        """Access pulse channel 1."""
        return self._pulse1

    @property
    def pulse2(self) -> PulseChannel:
        """Access pulse channel 2."""
        return self._pulse2

    @property
    def triangle(self) -> TriangleChannel:
        """Access triangle channel."""
        return self._triangle

    @property
    def noise(self) -> NoiseChannel:
        """Access noise channel."""
        return self._noise

File: familybox/apu/apu.c
/* APU: pulse x2, triangle, noise + frame counter + sample output */
#include "nes_internal.h"
#include <string.h>

static const int LENGTH_TABLE[32] = {10, 254, 20, 2,  40, 4,  80, 6,  160, 8,  60, 10, 14, 12, 26, 14,
                                     12, 16,  24, 18, 48, 20, 96, 22, 192, 24, 72, 26, 16, 28, 32, 30};

static const int DUTY_TABLE[4][8] = {
    {0, 1, 0, 0, 0, 0, 0, 0}, {0, 1, 1, 0, 0, 0, 0, 0}, {0, 1, 1, 1, 1, 0, 0, 0}, {1, 0, 0, 1, 1, 1, 1, 1}};

static const int TRI_TABLE[32] = {15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5,  4,  3,  2,  1,  0,
                                  0,  1,  2,  3,  4,  5,  6, 7, 8, 9, 10, 11, 12, 13, 14, 15};

static const int NOISE_PERIOD[16] = {4, 8, 16, 32, 64, 96, 128, 160, 202, 254, 380, 508, 762, 1016, 2034, 4068};

void apu_reset(Nes *n)
{
    Apu *a = &n->apu;
    memset(a, 0, sizeof(*a));
    a->noise.shift_reg = 1;
    a->sample_rate = 44100;
    a->cycles_per_sample = 1789773.0 / 44100.0;
    a->sample_timer = 0.0;
}

static void pulse_write(PulseCh *c, int reg, uint8_t v)
{
    switch (reg)
    {
    case 0:
        c->duty = (v >> 6) & 3;
        c->envelope_loop = (v & 0x20) != 0;
        c->constant_volume = (v & 0x10) != 0;
        c->volume = v & 0x0F;
        break;
    case 1:
        c->sweep_enabled = (v & 0x80) != 0;
        c->sweep_period = (v >> 4) & 7;
        c->sweep_negate = (v & 0x08) != 0;
        c->sweep_shift = v & 7;
        c->sweep_reload = 1;
        break;
    case 2:
        c->timer_period = (c->timer_period & 0x700) | v;
        break;
    case 3:
        c->timer_period = (c->timer_period & 0xFF) | ((v & 7) << 8);
        if (c->enabled)
            c->length_counter = LENGTH_TABLE[(v >> 3) & 0x1F];
        c->envelope_start = 1;
        c->duty_index = 0;
        break;
    }
}

static void pulse_quarter(PulseCh *c)
{
    if (c->envelope_start)
    {
        c->envelope_start = 0;
        c->envelope_decay = 15;
        c->envelope_divider = c->volume;
    }
    else if (c->envelope_divider > 0)
    {
        c->envelope_divider--;
    }
    else
    {
        c->envelope_divider = c->volume;
        if (c->envelope_decay > 0)
            c->envelope_decay--;
        else if (c->envelope_loop)
            c->envelope_decay = 15;
    }
}

static void pulse_half(PulseCh *c)
{
    if (!c->envelope_loop && c->length_counter > 0)
        c->length_counter--;
    if (c->sweep_divider == 0 && c->sweep_enabled && c->sweep_shift > 0 && c->timer_period >= 8)
    {
        int change = c->timer_period >> c->sweep_shift;
        if (c->sweep_negate)
            c->timer_period = c->timer_period - change;
        else
            c->timer_period = c->timer_period + change;
        /* Oversize periods mute the channel on real hardware; do not wrap. */
    }
    if (c->sweep_divider == 0 || c->sweep_reload)
    {
        c->sweep_divider = c->sweep_period;
        c->sweep_reload = 0;
    }
    else
    {
        c->sweep_divider--;
    }
}

static int pulse_out(PulseCh *c)
{
    if (!c->enabled || c->length_counter == 0)
        return 0;
    if (c->timer_period < 8 || c->timer_period > 0x7FF)
        return 0;
    int env = c->constant_volume ? c->volume : c->envelope_decay;
    return DUTY_TABLE[c->duty][c->duty_index] ? env : 0;
}

/* The waveform timers are clocked on the APU clock (CPU/2), so each timer
   step lasts (period + 1) * 2 CPU cycles. The channel state stores CPU
   cycles remaining until the next step fires. */
static void pulse_tick_timer(PulseCh *c, int cycles)
{
    int remaining = cycles;
    while (remaining > 0)
    {
        if (c->timer > remaining)
        {
            c->timer -= remaining;
            break;
        }
        remaining -= c->timer;
        c->timer = (c->timer_period + 1) * 2;
        c->duty_index = (c->duty_index + 1) & 7;
    }
}

static void tri_tick_timer(TriangleCh *t, int cycles)
{
    int remaining = cycles;
    while (remaining > 0)
    {
        if (t->timer > remaining)
        {
            t->timer -= remaining;
            break;
        }
        remaining -= t->timer;
        t->timer = (t->timer_period + 1) * 2;
        if (t->length_counter > 0 && t->linear_counter > 0)
        {
            t->sequence_index = (t->sequence_index + 1) & 31;
        }
    }
}

static void noise_tick_timer(NoiseCh *nch, int cycles)
{
    int remaining = cycles;
    while (remaining > 0)
    {
        if (nch->timer > remaining)
        {
            nch->timer -= remaining;
            break;
        }
        remaining -= nch->timer;
        nch->timer = (nch->timer_period + 1) * 2;
        int bit0 = nch->shift_reg & 1;
        int bit1 = (nch->shift_reg >> (nch->mode ? 6 : 1)) & 1;
        int fb = bit0 ^ bit1;
        nch->shift_reg = (uint16_t)((nch->shift_reg >> 1) | (fb << 14));
    }
}

static void tri_quarter(TriangleCh *t)
{
    if (t->linear_reload_flag)
        t->linear_counter = t->linear_counter_reload;
    else if (t->linear_counter > 0)
        t->linear_counter--;
    if (!t->linear_control)
        t->linear_reload_flag = 0;
}

static void tri_half(TriangleCh *t)
{
    if (!t->linear_control && t->length_counter > 0)
        t->length_counter--;
}

static int tri_out(TriangleCh *t)
{
    if (!t->enabled || t->length_counter == 0 || t->linear_counter == 0)
        return 0;
    return TRI_TABLE[t->sequence_index];
}

static void noise_quarter(NoiseCh *c)
{
    if (c->envelope_start)
    {
        c->envelope_start = 0;
        c->envelope_decay = 15;
        c->envelope_divider = c->volume;
    }
    else if (c->envelope_divider > 0)
    {
        c->envelope_divider--;
    }
    else
    {
        c->envelope_divider = c->volume;
        if (c->envelope_decay > 0)
            c->envelope_decay--;
        else if (c->envelope_loop)
            c->envelope_decay = 15;
    }
}

static void noise_half(NoiseCh *c)
{
    if (!c->envelope_loop && c->length_counter > 0)
        c->length_counter--;
}

static int noise_out(NoiseCh *c)
{
    if (!c->enabled || c->length_counter == 0)
        return 0;
    if (c->shift_reg & 1)
        return 0;
    return c->constant_volume ? c->volume : c->envelope_decay;
}

void apu_reg_write(Nes *n, uint16_t addr, uint8_t v)
{
    Apu *a = &n->apu;
    if (addr >= 0x4000 && addr <= 0x4003)
        pulse_write(&a->pulse1, addr & 3, v);
    else if (addr >= 0x4004 && addr <= 0x4007)
        pulse_write(&a->pulse2, addr & 3, v);
    else if (addr == 0x4008)
    {
        a->tri.linear_control = (v & 0x80) != 0;
        a->tri.linear_counter_reload = v & 0x7F;
    }
    else if (addr == 0x400A)
    {
        a->tri.timer_period = (a->tri.timer_period & 0x700) | v;
    }
    else if (addr == 0x400B)
    {
        a->tri.timer_period = (a->tri.timer_period & 0xFF) | ((v & 7) << 8);
        if (a->tri.enabled)
            a->tri.length_counter = LENGTH_TABLE[(v >> 3) & 0x1F];
        a->tri.linear_reload_flag = 1;
    }
    else if (addr == 0x400C)
    {
        a->noise.envelope_loop = (v & 0x20) != 0;
        a->noise.constant_volume = (v & 0x10) != 0;
        a->noise.volume = v & 0x0F;
    }
    else if (addr == 0x400E)
    {
        a->noise.mode = (v & 0x80) != 0;
        a->noise.timer_period = NOISE_PERIOD[v & 0x0F];
    }
    else if (addr == 0x400F)
    {
        if (a->noise.enabled)
            a->noise.length_counter = LENGTH_TABLE[(v >> 3) & 0x1F];
        a->noise.envelope_start = 1;
    }
    else if (addr == 0x4015)
    {
        a->pulse1.enabled = (v & 0x01) != 0;
        a->pulse2.enabled = (v & 0x02) != 0;
        a->tri.enabled = (v & 0x04) != 0;
        a->noise.enabled = (v & 0x08) != 0;
        if (!(v & 0x01))
            a->pulse1.length_counter = 0;
        if (!(v & 0x02))
            a->pulse2.length_counter = 0;
        if (!(v & 0x04))
            a->tri.length_counter = 0;
        if (!(v & 0x08))
            a->noise.length_counter = 0;
    }
    else if (addr == 0x4017)
    {
        a->frame_mode = (v >> 7) & 1;
        a->irq_inhibit = (v & 0x40) != 0;
        a->frame_counter = 0;
        a->frame_step = 0;
    }
}

uint8_t apu_reg_read(Nes *n, uint16_t addr)
{
    if (addr != 0x4015)
        return 0;
    Apu *a = &n->apu;
    uint8_t v = 0;
    if (a->pulse1.length_counter > 0)
        v |= 0x01;
    if (a->pulse2.length_counter > 0)
        v |= 0x02;
    if (a->tri.length_counter > 0)
        v |= 0x04;
    if (a->noise.length_counter > 0)
        v |= 0x08;
    return v;
}

static float mix_sample(int p1, int p2, int t, int n)
{
    float pulse_out = 0.0f;
    if (p1 + p2 > 0)
    {
        pulse_out = 95.88f / ((8128.0f / (p1 + p2)) + 100.0f);
    }
    float tnd_out = 0.0f;
    float tnd = t / 8227.0f + n / 12241.0f;
    if (tnd > 0.0f)
    {
        tnd_out = 159.79f / ((1.0f / tnd) + 100.0f);
    }
    return pulse_out + tnd_out;
}

static void frame_quarter(Apu *a)
{
    pulse_quarter(&a->pulse1);
    pulse_quarter(&a->pulse2);
    tri_quarter(&a->tri);
    noise_quarter(&a->noise);
}

static void frame_half(Apu *a)
{
    pulse_half(&a->pulse1);
    pulse_half(&a->pulse2);
    tri_half(&a->tri);
    noise_half(&a->noise);
}

void apu_tick(Nes *n, int cpu_cycles, int16_t *pcm, int max_samples, int *out_count)
{
    Apu *a = &n->apu;
    *out_count = 0;

    a->frame_counter += cpu_cycles;
    int *steps4 = (int[]){7457, 14913, 22371, 29829};
    int *steps5 = (int[]){7457, 14913, 22371, 29829, 37281};
    int *steps = a->frame_mode ? steps5 : steps4;
    int max_steps = a->frame_mode ? 5 : 4;

    while (a->frame_step < max_steps && a->frame_counter >= steps[a->frame_step])
    {
        int half = (a->frame_mode == 0 && (a->frame_step == 1 || a->frame_step == 3)) ||
                   (a->frame_mode == 1 && (a->frame_step == 0 || a->frame_step == 2));
        frame_quarter(a);
        if (half)
            frame_half(a);
        a->frame_step++;
        if (a->frame_step >= max_steps)
        {
            a->frame_counter = 0;
            a->frame_step = 0;
        }
    }

    pulse_tick_timer(&a->pulse1, cpu_cycles);
    pulse_tick_timer(&a->pulse2, cpu_cycles);
    tri_tick_timer(&a->tri, cpu_cycles);
    noise_tick_timer(&a->noise, cpu_cycles);

    /* Produce ALL samples owed for these cycles */
    a->sample_timer += cpu_cycles;
    while (a->sample_timer >= a->cycles_per_sample)
    {
        a->sample_timer -= a->cycles_per_sample;
        if (!pcm || *out_count >= max_samples)
        {
            /* drop if no buffer space, but keep timer accurate */
            continue;
        }
        int p1 = pulse_out(&a->pulse1);
        int p2 = pulse_out(&a->pulse2);
        int t = tri_out(&a->tri);
        int nz = noise_out(&a->noise);
        int v = 0;
        if (p1 || p2 || t || nz)
        {
            float s = mix_sample(p1, p2, t, nz);
            float bipolar = s * 2.0f - 0.12f;
            v = (int)(bipolar * 28000.0f);
            if (v > 32767)
                v = 32767;
            if (v < -32768)
                v = -32768;
        }
        pcm[(*out_count)++] = (int16_t)v;
    }
}
